src/main/java/org/opensearch/ad/ExecuteADResultResponseRecorder.java

```diff
@@ -0,0 +1,268 @@
+    private Client client;
+
+    public ExecuteADResultResponseRecorder(
+        AnomalyDetectionIndices anomalyDetectionIndices,
+        AnomalyIndexHandler<AnomalyResult> anomalyResultHandler,
+        ADTaskManager adTaskManager,
+        DiscoveryNodeFilterer nodeFilter,
+        ThreadPool threadPool,
+        Client client
+    ) {
+        this.anomalyDetectionIndices = anomalyDetectionIndices;
+        this.anomalyResultHandler = anomalyResultHandler;
+        this.adTaskManager = adTaskManager;
+        this.nodeFilter = nodeFilter;
+        this.threadPool = threadPool;
+        this.client = client;
+    }
+
+    public void indexAnomalyResult(
+        Instant detectionStartTime,
+        Instant executionStartTime,
+        AnomalyResultResponse response,
+        AnomalyDetector detector
+    ) {
+        String detectorId = detector.getDetectorId();
+        try {
+            // skipping writing to the result index if not necessary
+            // For a single-entity detector, the result is not useful if error is null
+            // and rcf score (thus anomaly grade/confidence) is null.
+            // For a HCAD detector, we don't need to save on the detector level.
+            // We return 0 or Double.NaN rcf score if there is no error.
+            if ((response.getAnomalyScore() <= 0 || Double.isNaN(response.getAnomalyScore())) && response.getError() == null) {
+                updateRealtimeTask(response, detectorId);
+                return;
+            }
+            IntervalTimeConfiguration windowDelay = (IntervalTimeConfiguration) detector.getWindowDelay();
+            Instant dataStartTime = detectionStartTime.minus(windowDelay.getInterval(), windowDelay.getUnit());
+            Instant dataEndTime = executionStartTime.minus(windowDelay.getInterval(), windowDelay.getUnit());
+            User user = detector.getUser();
+
+            if (response.getError() != null) {
+                log.info("Anomaly result action run successfully for {} with error {}", detectorId, response.getError());
+            }
+
+            AnomalyResult anomalyResult = response
+                .toAnomalyResult(
+                    detectorId,
+                    dataStartTime,
+                    dataEndTime,
+                    executionStartTime,
+                    Instant.now(),
+                    anomalyDetectionIndices.getSchemaVersion(ADIndex.RESULT),
+                    user,
+                    response.getError()
+                );
+
+            String resultIndex = detector.getResultIndex();
+            anomalyResultHandler.index(anomalyResult, detectorId, resultIndex);
+            updateRealtimeTask(response, detectorId);
+        } catch (EndRunException e) {
+            throw e;
+        } catch (Exception e) {
+            log.error("Failed to index anomaly result for " + detectorId, e);
+        }
+    }
+
+    private void updateRealtimeTask(AnomalyResultResponse response, String detectorId) {
```

Comment:
added comment:

Update real time task (one document per detector in state index). If the real-time task has no changes compared with local cache, the task won't update. Task only updates when the state changed, or any error happened, or AD job stopped. Task is mainly consumed by the front-end to track detector status. For single-stream detectors, we embed model total updates in AnomalyResultResponse and update state accordingly. For HCAD, we won't wait for model finishing updating before returning a response to the job scheduler since it might be long before all entities finish execution. So we don't embed model total updates in AnomalyResultResponse. Instead, we issue a profile request to poll each model node and get the maximum total updates among all models.

Reply:
Thank you